### ISSUE TITLE
Error while updating terraform-created Azure SQL Server with unexpected status 400

### ISSUE THREAD

**Issue by PayelGGithub** [hashicorp/terraform-provider-azurerm#24390]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.6.2

### AzureRM Provider Version

3.80.0

### Affected Resource(s)/Data Source(s)

azurerm_mssql_server

### Terraform Configuration Files

```hcl
resource "azurerm_mssql_server" "sqlserver" {
  name                              = var.sqlservername
  resource_group_name               = data.azurerm_resource_group.sqldbrg.name
  location                          = data.azurerm_resource_group.sqldbrg.location
  version                           = "12.0"
  administrator_login               = data.azurerm_key_vault_secret.sqladminuser.value
  administrator_login_password      = data.azurerm_key_vault_secret.sqladminpwd.value
  public_network_access_enabled     = var.sqlserver_isPublic
  minimum_tls_version               = "1.2"
  primary_user_assigned_identity_id = data.azurerm_user_assigned_identity.sql_id.id

  dynamic "identity" {
    for_each = var.identity_type != "SystemAssigned" ? [1] : []
    content {
      type         = var.identity_type != "SystemAssigned" ? var.identity_type : "SystemAssigned"
      identity_ids = var.identity_type != "SystemAssigned" ? [data.azurerm_user_assigned_identity.sql_id.id] : null
    }
  }
  azuread_administrator {
    azuread_authentication_only = var.azuread_authentication_only
    login_username              = "AzureAD Admin"
    object_id                   = var.sqlad_adminid
  }
  tags = var.tags

  lifecycle {
    ignore_changes = [transparent_data_encryption_key_vault_key_id]
  }
}

resource "azurerm_mssql_server_transparent_data_encryption" "this" {
  count                 = var.enable_mssql_server_transparent_data_encryption ? 1 : 0
  server_id             = azurerm_mssql_server.sqlserver.id
  key_vault_key_id      = data.azurerm_key_vault_key.this.id
  auto_rotation_enabled = true
}


## SQL DB ##
resource "azurerm_mssql_database" "sqldb" {
  name                        = var.sqldbname
  server_id                   = azurerm_mssql_server.sqlserver.id
  collation                   = "SQL_Latin1_General_CP1_CI_AS" 
  min_capacity                = var.sqldb_mincapacity
  auto_pause_delay_in_minutes = -1
  read_scale                  = var.sqldb_readscale
  sku_name                    = var.sqlskuname
  max_size_gb                 = var.maxsizegb
  zone_redundant              = var.zone_redundant
  geo_backup_enabled          = var.geobackuppolicy
  storage_account_type        = var.bkpstoragetype
  tags                        = var.tags
  long_term_retention_policy {
    weekly_retention  = var.ltrpolicy_weeklyretention
    monthly_retention = var.ltrpolicy_monthlyretention
    yearly_retention  = var.ltrpolicy_yearlyretention
    week_of_year      = var.ltrpolicy_weekofyear
  }
  short_term_retention_policy {
    retention_days           = var.strpolicy_retentiondays
    backup_interval_in_hours = var.strpolicy_backupinterval
  }
```

### Debug Output/Panic Output

```shell
https://gist.github.com/PayelGGithub/bf5934904cd75df1ee89fa1947ba1436
```

### Expected Behaviour

mssql server and database module ran successfully without any issue.

### Actual Behaviour

│ Error: updating Server (Subscription: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
│ Resource Group Name: "rg-metadata-xxxx-sbx-qc-001"
│ Server Name: "asql-xxxx-sbx-qc-88"): performing CreateOrUpdate: unexpected status 400 with error: InvalidResource: The resource definition is invalid.
│ 
│   with azurerm_mssql_server.sqlserver,
│   on main.tf line 59, in resource "azurerm_mssql_server" "sqlserver":
│   59: resource "azurerm_mssql_server" "sqlserver" {

### Steps to Reproduce

terraform plan
terraform apply

### Important Factoids

Running in Qatar Central location. have created Azure SQL Server module and deployed it along with database with Terraform via Azure DevOps pipeline and it worked absolutely fine in the first run. When I am rerunning the pipeline without even changing the code, it throws me the above error.

**Comment by sinbai** [comment]
Thanks for opening this issue. In order to find the root cause, a Terraform configuration file (containing dependent resources and variable values) and detailed steps would be of great help in quickly reproducing/troubleshooting the issue. Could you please provide the complete Terraform configuration and detailed repro steps to help reproduce this issue? Thanks in advanced.

In addition, based on the information provided, it is recommended to compare the differences between the first and second request bodies in order to locate specific attributes to find and solve problems. BTW, configuring [TF Log](https://developer.hashicorp.com/terraform/internals/debugging) is one of the ways to get the data of the two requests.

**Comment by PayelGGithub** [comment]
Hello Elena,

Thanks for your response. Please find the configuration file attached herewith this email thread.

Regards,
Payel

**Comment by harshavmb** [comment]
Hi @PayelGGithub ,

Regarding below comment.. 

> Running in Qatar Central location. have created Azure SQL Server module and deployed it along with database with Terraform via Azure DevOps pipeline and it worked absolutely fine in the first run. When I am rerunning the pipeline without even changing the code, it throws me the above error.

For the debug logs you shared, it failed while making a `PUT` request. This won't get triggered unless there is a change in the configuration. Can you post output of `terraform plan` on subsequent runs?

Note:: It's recommended to redact the metadata such as resourceId, subscriptionId, tenantId information when posting here/gist, otherwise you are exposing your tenant to internet.

**Comment by PayelGGithub** [comment]
Hello, I have shared the terraform plan logs in the new gist link here:

https://gist.github.com/PayelGGithub/c9e32380529b2515fca3e5975666c288

Please note, I have redact the required metadata like resourceId, subscriptionId, tenantId information.

**Comment by harshavmb** [comment]
So, from the plan you shared it's evident that there is a drift in the tags & this is forcing an update of the resource. I couldn't replicate the issue at the end. 

To ignore the drift, you could append your `lifecycle` block to ignore `tags`. 

```bash
lifecycle {
    ignore_changes = [transparent_data_encryption_key_vault_key_id]
  }
```

As an alternate you could try with newer versions of `azurerm` provider to see if the issue is still persistent.

**Comment by wozorio** [comment]
I am facing the same issue with the `azurerm` provider version `1.87.0` when trying to change the SQL server `connection_policy` to `Redirect`.

Terraform Plan shows that the connection policy will be updated as intended, but Terraform Apply fails with:
```bash
performing CreateOrUpdate: unexpected status 400 with error: InvalidResource: The resource definition is invalid.
```

**Comment by deckerbr1979** [comment]
I am experiencing the same issue when trying to update a SQL Server deployed into Azure.  Any updates on the resolution for this issue?

> hashicorp/azurerm v3.89.0

**TF Apply:**

```
~ update in-place

Terraform will perform the following actions:

 # module.azurerm_mssql_server.azurerm_mssql_server.mssqls will be updated in-place
  ~ resource "azurerm_mssql_server" "mssqls" {
        id                                   = "/subscriptions/***/resourceGroups/rg-app0-cus-dev-001/providers/Microsoft.Sql/servers/mssqls-app0-cus-dev-001"
        name                                 = "mssqls-app0-cus-dev-001"
      ~ outbound_network_restriction_enabled = false -> true
      ~ public_network_access_enabled        = true -> false
        # (9 unchanged attributes hidden)

        # (2 unchanged blocks hidden)
    }

```

**TF Apply Results:**

```
│ Error: updating Server (Subscription: "***"
│ Resource Group Name: "rg-app0-cus-dev-001"
│ Server Name: "mssqls-app0-cus-dev-001"): performing CreateOrUpdate: unexpected status 400 with error: InvalidResource: The resource definition is invalid.
│ 
│   with module.azurerm_mssql_server.azurerm_mssql_server.mssqls,
│   on .terraform/modules/azurerm_mssql_server/main.tf line 20, in resource "azurerm_mssql_server" "mssqls":
│   20: resource "azurerm_mssql_server" "mssqls" {
```

The resource can be successfully removed from the statefile and successfully imported  back into the statefile.

If the changes are made manually within the portal they are successfully and TF does not try and update the resource.

**Comment by bertsch-ronja-office** [comment]
I am facing a similar issue when setting the Entra ID admin to a service principal:

  azuread_administrator = {
    login_username = <<SP-Displayname>>
    object_id      = <<SP-ObjectId>>
    tenant_id      = data.azurerm_client_config.current.tenant_id
  }

**Comment by PayelGGithub** [comment]
We deleted the state file from backend storage account for mssql and redeployed the resource. On rerun of it, there was no further issue. Not sure, if this can resolve other's issue, but it worked for us.

**Comment by bertsch-ronja-office** [comment]
@PayelGGithub Your approach might work for empty resources, but we cannot just redeploy an existing SQL server that has multiple databases attached.

**Comment by joshvito** [comment]
I also had this error of 400 bad request present itself on Apply when a Plan was okay.

**tldr**: I fixed my issue by setting `primary_user_assigned_identity_id` and `identity_ids` in the same update. If your add the ids in one step and try to set the primary user identity in a follow up step, you will get a 400 Bad Request as detailed below.

- I also[ put in a PR ](https://github.com/hashicorp/terraform-provider-azurerm/pull/25535)to update the documentation, for clarity.

Here is an example of my configuration file:
```
resource "azurerm_mssql_server" "this" {
  name                = local.dbServerName
  resource_group_name = var.this_resource_group_name
  location            = var.location

  version                      = "12.0"
  minimum_tls_version          = "1.2"
  administrator_login          = random_pet.admin_login.id
  administrator_login_password = random_password.admin_password.result

  public_network_access_enabled = true

  identity {
    type = "SystemAssigned, UserAssigned"
    identity_ids = [
      data.azurerm_user_assigned_identity.this.id
    ]
  }
  # this is the one property that was added to the resource
  primary_user_assigned_identity_id = data.azurerm_user_assigned_identity.this.id

  azuread_administrator {
    login_username = data.azuread_group.azuread_administrators.display_name
    object_id      = data.azuread_group.azuread_administrators.object_id
  }

  tags = local.resource_tags
}
```


Resulted in the following PLAN output
![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/1249866/9fc5a170-c97f-4aac-a040-32d39dea1934)


## Logs
Here are the logs from the terraform apply command:

```
Content-Length: 0
Content-Type: application/json; charset=utf-8
X-Ms-Correlation-Request-Id: <guid>
Accept-Encoding: gzip: timestamp=2024-04-08T15:12:17.894Z
2024-04-08T15:12:17.894Z [DEBUG] provider.terraform-provider-azurerm_v3.98.0_x5: PUT https://management.azure.com/subscriptions/<subscription_guid>/resourceGroups/<rg_name>/providers/Microsoft.Sql/servers/<server_name>?api-version=2023-02-01-preview: timestamp=2024-04-08T15:12:17.894Z
2024-04-08T15:12:17.959Z [DEBUG] provider.terraform-provider-azurerm_v3.98.0_x5: AzureRM Response for https://management.azure.com/subscriptions/<subscription_guid>/resourceGroups/<rg_name>/providers/Microsoft.Sql/servers/<server_name>?api-version=2023-02-01-preview: 
HTTP/2.0 400 Bad Request
Content-Length: 84
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
Date: Mon, 08 Apr 2024 15:12:17 GMT
Expires: -1
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Cache: CONFIG_NOCACHE
X-Content-Type-Options: nosniff
X-Ms-Correlation-Request-Id: <guid>
X-Ms-Failure-Cause: gateway
X-Ms-Request-Id: <guid>
X-Ms-Routing-Request-Id: SOUTHCENTRALUS:20240408T151217Z:<guid>
X-Msedge-Ref: Ref A: AE6931ECFF4D4BD59592029C7EE79E98 Ref B: SN4AA2022305053 Ref C: 2024-04-08T15:12:17Z

{"error":{"code":"InvalidResource","message":"The resource definition is invalid."}}: timestamp=2024-04-08T15:12:17.959Z
2024-04-08T15:12:17.959Z [TRACE] provider.terraform-provider-azurerm_v3.98.0_x5: Called downstream: tf_resource_type=azurerm_mssql_server @module=sdk.helper_schema tf_provider_addr=provider tf_req_id=<guid> @caller=github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:920 tf_rpc=ApplyResourceChange timestamp=2024-04-08T15:12:17.959Z
2024-04-08T15:12:17.960Z [TRACE] provider.terraform-provider-azurerm_v3.98.0_x5: Received downstream response: diagnostic_error_count=1 diagnostic_warning_count=0 tf_proto_version=5.4 tf_req_id=<guid> tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/tf5serverlogging/downstream_request.go:40 @module=sdk.proto tf_provider_addr=provider tf_req_duration_ms=551 tf_resource_type=azurerm_mssql_server timestamp=2024-04-08T15:12:17.960Z
2024-04-08T15:12:17.960Z [ERROR] provider.terraform-provider-azurerm_v3.98.0_x5: Response contains error diagnostic: tf_provider_addr=provider tf_req_id=<guid> tf_rpc=ApplyResourceChange diagnostic_summary="updating Server (Subscription: "<subscription_guid>"
Resource Group Name: "<rg_name>"
Server Name: "<server_name>"): performing CreateOrUpdate: unexpected status 400 (400 Bad Request) with error: InvalidResource: The resource definition is invalid." tf_proto_version=5.4 diagnostic_severity=ERROR tf_resource_type=azurerm_mssql_server @caller=github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/diag/diagnostics.go:58 @module=sdk.proto diagnostic_detail= timestamp=2024-04-08T15:12:17.960Z
2024-04-08T15:12:17.960Z [TRACE] provider.terraform-provider-azurerm_v3.98.0_x5: Served request: tf_req_id=<guid> tf_resource_type=azurerm_mssql_server @caller=github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/tf5server/server.go:872 @module=sdk.proto tf_proto_version=5.4 tf_provider_addr=provider tf_rpc=ApplyResourceChange timestamp=2024-04-08T15:12:17.960Z
2024-04-08T15:12:17.961Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for azurerm_mssql_server.this
2024-04-08T15:12:17.961Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for azurerm_mssql_server.this
2024-04-08T15:12:17.961Z [TRACE] evalApplyProvisioners: azurerm_mssql_server.this is not freshly-created, so no provisioning is required
2024-04-08T15:12:17.961Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for azurerm_mssql_server.this
2024-04-08T15:12:17.961Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for azurerm_mssql_server.this
2024-04-08T15:12:17.961Z [DEBUG] State storage *remote.State declined to persist a state snapshot
2024-04-08T15:12:17.961Z [ERROR] vertex "azurerm_mssql_server.this" error: updating Server (Subscription: "<subscription_guid>"
Resource Group Name: "<rg_name>"
Server Name: "<server_name>"): performing CreateOrUpdate: unexpected status 400 (400 Bad Request) with error: InvalidResource: The resource definition is invalid.
2024-04-08T15:12:17.961Z [TRACE] vertex "azurerm_mssql_server.this": visit complete, with errors
2024-04-08T15:12:17.961Z [TRACE] dag/walk: upstream of "output.this_mssql_server_name (expand)" errored, so skipping
2024-04-08T15:12:17.961Z [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" errored, so skipping
2024-04-08T15:12:17.961Z [TRACE] dag/walk: upstream of "output.this_mssql_server_id (expand)" errored, so skipping
2024-04-08T15:12:17.961Z [TRACE] dag/walk: upstream of "root" errored, so skipping
2024-04-08T15:12:17.961Z [DEBUG] states/remote: state read serial is: 12; serial is: 12
2024-04-08T15:12:17.961Z [DEBUG] states/remote: state read lineage is: 4571ddfd-e437-6c71-219a-043e510a1dc2; lineage is: 4571ddfd-e437-6c71-219a-043e510a1dc2
╷
│ Error: updating Server (Subscription: "<subscription_guid>"
│ Resource Group Name: "<rg_name>"
│ Server Name: "<server_name>"): performing CreateOrUpdate: unexpected status 400 (400 Bad Request) with error: InvalidResource: The resource definition is invalid.
│ 
│   with azurerm_mssql_server.this,
│   on main.tf line 110, in resource "azurerm_mssql_server" "this":
│  110: resource "azurerm_mssql_server" "this" {
│ 
╵

```
